README.md
[![pub package](https://img.shields.io/pub/v/flutter_native_admob.svg)](https://pub.dartlang.org/packages/flutter_native_admob)

# flutter_native_admob

Plugin to integrate Firebase Native Admob to Flutter application
Platform supported: **iOS**, **Android**

# Getting Started

For help getting started with Flutter, view our online [documentation](http://flutter.io/).

## Setup Android project

1. Add the classpath to the [project]/android/build.gradle file.

```s
dependencies {
  // Example existing classpath
  classpath 'com.android.tools.build:gradle:3.2.1'
  // Add the google services classpath
  classpath 'com.google.gms:google-services:4.3.0'
}
```

2. Add the apply plugin to the [project]/android/app/build.gradle file.

```s
// ADD THIS AT THE BOTTOM
apply plugin: 'com.google.gms.google-services'
```

3. Add your Admob App ID.

**_Important_**: This step is required as of Google Mobile Ads SDK version 17.0.0. Failure to add this <meta-data> tag results in a crash with the message: The Google Mobile Ads SDK was initialized incorrectly.

```xml
<manifest>
  <application>
    <!-- Sample AdMob App ID: ca-app-pub-3940256099942544~3347511713 -->
    <meta-data
      android:name="com.google.android.gms.ads.APPLICATION_ID"
      android:value="ca-app-pub-xxxxxxxxxxxxxxxx~yyyyyyyyyy"/>
  </application>
</manifest>
```

## Setup iOS project

1. Add Admob App ID:

**_Important_**: This step is required as of Google Mobile Ads SDK version 7.42.0. Failure to add add this Info.plist entry results in a crash with the message: The Google Mobile Ads SDK was initialized incorrectly.

In your app's Info.plist file, add a GADApplicationIdentifier key with a string value of your AdMob app ID. You can find your App ID in the AdMob UI.

You can make this change programmatically:

```xml
<key>GADApplicationIdentifier</key>
<string>Your_Admob_App_ID</string>
```

2. Add embeded view support:

In your app's Info.plist file, add this

```xml
<key>io.flutter.embedded_views_preview</key>
<true/>
```

## How it works

`NativeAdmob` is a Flutter widget, so you can add it anywhere in Flutter application.

| Property   | Description                                       | Type                  |
| :--------- | :------------------------------------------------ | :-------------------- |
| adUnitID   | Your ad unit ID to load                           | String                |
| numberAds  | Number of ads to load                             | Int                   |
| loading    | A widget to show when the ad is loading           | Widget                |
| error      | A widget to show when the ad got error            | Widget                |
| options    | Native ad styling options                         | NativeAdmobOptions    |
| type       | Native ad type (banner or full)                   | NativeAdmobType.full  |
| controller | Controller for controlling the NativeAdmob widget | NativeAdmobController |

#### `NativeAdmobOptions`

| Property            | Description                                                                                    | Type            | Default value                                                           |
| ------------------- | ---------------------------------------------------------------------------------------------- | --------------- | ----------------------------------------------------------------------- |
| showMediaContent    | Whether to show the media content or not                                                       | bool            | true                                                                    |
| ratingColor         | Rating star color                                                                              | Color           | Colors.yellow                                                           |
| adLabelTextStyle    | The ad label on the top left corner                                                            | NativeTextStyle | `fontSize: 12, color: Colors.white, backgroundColor: Color(0xFFFFCC66)` |
| headlineTextStyle   | The ad headline title                                                                          | NativeTextStyle | `fontSize: 16, color: Colors.black`                                     |
| advertiserTextStyle | Identifies the advertiser. For example, the advertiser’s name or visible URL. (below headline) | NativeTextStyle | `fontSize: 14, color: Colors.black`                                     |
| bodyTextStyle       | The ad description                                                                             | NativeTextStyle | `fontSize: 12, color: Colors.grey`                                      |
| storeTextStyle      | The app store name. For example, "App Store".                                                  | NativeTextStyle | `fontSize: 12, color: Colors.black`                                     |
| priceTextStyle      | String representation of the app's price.                                                      | NativeTextStyle | `fontSize: 12, color: Colors.black`                                     |
| callToActionStyle   | Text that encourages user to take some action with the ad. For example "Install".              | NativeTextStyle | `fontSize: 15, color: Colors.white, backgroundColor: Color(0xFF4CBE99)` |

#### `NativeTextStyle`

| Property        | Description            | Type   |
| :-------------- | :--------------------- | :----- |
| fontSize        | Text font size         | double |
| color           | Text color             | Color  |
| backgroundColor | Background color       | Color  |
| isVisible       | Whether to show or not | bool   |

#### `NativeAdmobController`

| Property/Function                                             | Description                                                                         | Type                |
| :------------------------------------------------------------ | :---------------------------------------------------------------------------------- | :------------------ |
| stateChanged                                                  | Stream that notify each time the loading state changed                              | Stream<AdLoadState> |
| void setAdUnitID(String adUnitID)                             | Change the ad unit ID, it will load the ad again if the id is changed from previous |                     |
| void reloadAd({bool forceRefresh = false, int numberAds = 1}) | Reload the ad with optionals number of ads                                          |                     |
| void setTestDeviceIds(List<String> ids)                       | Add your test devices                                                               |                     |
| void setNonPersonalizedAds(bool nonPersAds)                   | Set the option to disable the personalized Ads. AdMob default option: personalized  |                     |

## Examples

### Default

```dart
NativeAdmob(
  adUnitID: "<Your ad unit ID>"
)
```

### Using controller, loading, error widget, type and options

```dart
final _controller = NativeAdmobController();

// Optional: enter your test device ids here
_controller.setTestDeviceIds([
  "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
  "YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY"
]);

// Optional: Set the option to disable the personalized Ads. AdMob default option: personalized
_controller.setNonPersonalizedAds(true);

NativeAdmob(
  adUnitID: "<Your ad unit ID>",
  loading: Center(child: CircularProgressIndicator()),
  error: Text("Failed to load the ad"),
  controller: _controller,
  type: NativeAdmobType.full,
  options: NativeAdmobOptions(
    ratingColor: Colors.red,
    // Others ...
  ),
)
```

### Hide the ad until it load completed

```dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_native_admob/flutter_native_admob.dart';
import 'package:flutter_native_admob/native_admob_controller.dart';

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  static const _adUnitID = "<Your ad unit ID>";

  final _nativeAdController = NativeAdmobController();
  double _height = 0;

  StreamSubscription _subscription;

  @override
  void initState() {
    _subscription = _nativeAdController.stateChanged.listen(_onStateChanged);
    super.initState();
  }

  @override
  void dispose() {
    _subscription.cancel();
    _nativeAdController.dispose();
    super.dispose();
  }

  void _onStateChanged(AdLoadState state) {
    switch (state) {
      case AdLoadState.loading:
        setState(() {
          _height = 0;
        });
        break;

      case AdLoadState.loadCompleted:
        setState(() {
          _height = 330;
        });
        break;

      default:
        break;
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: ListView(
          children: <Widget>[
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              height: _height,
              padding: EdgeInsets.all(10),
              margin: EdgeInsets.only(bottom: 20.0),
              child: NativeAdmob(
                // Your ad unit id
                adUnitID: _adUnitID,
                controller: _nativeAdController,

                // Don't show loading widget when in loading state
                loading: Container(),
              ),
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
          ],
        ),
      ),
    );
  }
}
```

### Prevent ad from reloading on ListView/GridView

When putting `NativeAdmob` in ListView/GridView, it will keep reloading as the `PlatformView` init again when scrolling to the item. To prevent from reloading and take full control of the `NativeAdmob`, we can create `NativeAdmobController` and keep it

```dart
import 'package:flutter/material.dart';
import 'package:flutter_native_admob/flutter_native_admob.dart';
import 'package:flutter_native_admob/native_admob_controller.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  static const _adUnitID = "<Your ad unit ID>";

  final _controller = NativeAdmobController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: ListView(
          children: <Widget>[
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              height: 330,
              padding: EdgeInsets.all(10),
              margin: EdgeInsets.only(bottom: 20.0),
              child: NativeAdmob(
                adUnitID: _adUnitID,
                controller: _controller,
              ),
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              margin: EdgeInsets.only(bottom: 20.0),
              height: 200.0,
              color: Colors.green,
            ),
            Container(
              height: 330,
              padding: EdgeInsets.all(10),
              margin: EdgeInsets.only(bottom: 20.0),
              child: NativeAdmob(
                adUnitID: _adUnitID,
                controller: _controller,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 0.0.1

- Support for Android only

## 1.0.0

- Support for both Android and iOS

## 1.0.1

- Remove BannerStyle
- Add BannerOptions to allow styling better

## 2.0.0

Major updates

- Change widget to `NativeAdmob`
- Add new `NativeAdmobOptions` for native styling
- Add `NativeAdmobController` for fully control the widget

## 2.0.0+2

Fix iOS and Android layout and some elements that is not applied for the style

## 2.0.0+3

Allow to dispose admob controller for better memory

## 2.1.0

- Allow to show/hide in `NativeTextStyle`
- Add `NativeAdmobType` that allows to change the ad layout

## 2.1.0+1

Add setTestDeviceIds() to NativeAdmobController

## 2.1.0+2

Add setNonPersonalizedAds() to NativeAdmobController

## 2.1.0+3

Allow to set number of ads when calling loadAds on NativeAdmobController

## 2.2.0

Null Safety support


lib/native_admob_options.dart
import 'package:flutter/material.dart';

class NativeTextStyle {
  final double? fontSize;
  final Color? color;
  final Color? backgroundColor;
  final bool isVisible;

  const NativeTextStyle({
    this.fontSize,
    this.color,
    this.backgroundColor,
    this.isVisible = true,
  });

  Map<String, dynamic> toJson() => {
        "backgroundColor": backgroundColor != null ? "#${backgroundColor!.value.toRadixString(16)}" : null,
        "fontSize": fontSize,
        "color": color != null ? "#${color!.value.toRadixString(16)}" : null,
        "isVisible": isVisible,
      };
}

class NativeAdmobOptions {
  final bool showMediaContent;
  final Color ratingColor;
  final Color? backgroundColor;
  final NativeTextStyle adLabelTextStyle;
  final NativeTextStyle headlineTextStyle;
  final NativeTextStyle advertiserTextStyle;
  final NativeTextStyle bodyTextStyle;
  final NativeTextStyle storeTextStyle;
  final NativeTextStyle priceTextStyle;
  final NativeTextStyle callToActionStyle;

  const NativeAdmobOptions({
    this.showMediaContent = true,
    this.ratingColor = Colors.yellow,
    this.backgroundColor,
    this.adLabelTextStyle = const NativeTextStyle(
      fontSize: 12,
      color: Colors.white,
      backgroundColor: Color(0xFFFFCC66),
    ),
    this.headlineTextStyle = const NativeTextStyle(
      fontSize: 16,
      color: Colors.black,
    ),
    this.advertiserTextStyle = const NativeTextStyle(
      fontSize: 14,
      color: Colors.black,
    ),
    this.bodyTextStyle = const NativeTextStyle(
      fontSize: 12,
      color: Colors.grey,
    ),
    this.storeTextStyle = const NativeTextStyle(
      fontSize: 12,
      color: Colors.black,
    ),
    this.priceTextStyle = const NativeTextStyle(
      fontSize: 12,
      color: Colors.black,
    ),
    this.callToActionStyle = const NativeTextStyle(
      fontSize: 15,
      color: Colors.white,
      backgroundColor: Color(0xFF4CBE99),
    ),
  });

  Map<String, dynamic> toJson() => {
        "showMediaContent": this.showMediaContent,
        "ratingColor": "#${ratingColor.value.toRadixString(16)}",
        "backgroundColor": backgroundColor != null ? "#${backgroundColor!.value.toRadixString(16)}" : null,
        "adLabelTextStyle": adLabelTextStyle.toJson(),
        "headlineTextStyle": headlineTextStyle.toJson(),
        "advertiserTextStyle": advertiserTextStyle.toJson(),
        "bodyTextStyle": bodyTextStyle.toJson(),
        "storeTextStyle": storeTextStyle.toJson(),
        "priceTextStyle": priceTextStyle.toJson(),
        "callToActionStyle": callToActionStyle.toJson(),
      };
}
